train.py

```diff
@@ -202,18 +114,20 @@ def main(conf: DictConfig) -> None:
         monitor=monitor_metric, min_delta=0.00, patience=18, mode=mode
     )
 
-    trainer_args = cast(dict[str, Any], OmegaConf.to_object(conf.trainer))
-
-    trainer_args["callbacks"] = [checkpoint_callback, early_stopping_callback]
-    trainer_args["logger"] = [tb_logger, csv_logger]
-    trainer_args["default_root_dir"] = experiment_dir
-    trainer = Trainer(**trainer_args)
+    # Define trainer
+    trainer: Trainer = instantiate(
+        conf.trainer,
+        callbacks=[checkpoint_callback, early_stopping_callback],
+        logger=[tb_logger, csv_logger],
+        default_root_dir=experiment_dir,
+    )
 
-    ######################################
-    # Run experiment
-    ######################################
+    # Train
     trainer.fit(model=task, datamodule=datamodule)
-    trainer.test(ckpt_path="best", datamodule=datamodule)
+
+    # Test
+    if hasattr(task, "test_step") and hasattr(datamodule, "test_dataloader"):
+        trainer.test(ckpt_path="best", datamodule=datamodule)
```

Comment:
Unfortunately, all trainers have a test_step (might be a no-op) and all datamodules have a test_dataloader (might raise an error). I would instead use:
```suggestion
    try:
        trainer.test(ckpt_path="best", datamodule=datamodule)
    except MisconfigurationException:
        pass
```
Can add more exceptions if other errors might be raised

Reply:
We could also check if test_dataset=None as well but yeah I think your solution works best.

Reply:
That won't work either because it could be using `self.dataset` instead of `self.test_dataset`.